### PR TITLE
feat: shuffle option in list api

### PR DESCRIPTION
**Pull request by wingkwong**
## Change Summary

ref: #1432 

- add shuffle option (0 / 1) - Default: 0 for listing
- if it is set to 1, then the order will be shuffled
- add alias `r` as `s` is in use for `sort`
- add shuffle parma for swagger ui
- add related info in docs

## Change type

- [x] feat: (new feature for the user, not a new feature for build script)
- [ ] fix: (bug fix for the user, not a fix to a build script)
- [x] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [ ] chore: (updating grunt tasks etc; no production code change)

## Test/ Verification

Base Data 

image

Without Shuffling - Test in Swagger UI (shuffle = 0)

Result:

```json
{
  "list": [
    {
      "Id": 1,
      "Title": "A",
      "CreatedAt": "2022-07-05 15:45:09",
      "UpdatedAt": "2022-07-05 15:45:09"
    },
    {
      "Id": 2,
      "Title": "B",
      "CreatedAt": "2022-07-05 15:45:10",
      "UpdatedAt": "2022-07-05 15:45:10"
    },
    {
      "Id": 3,
      "Title": "C",
      "CreatedAt": "2022-07-05 15:54:41",
      "UpdatedAt": "2022-07-05 15:54:41"
    },
    {
      "Id": 4,
      "Title": "D",
      "CreatedAt": "2022-07-05 15:54:42",
      "UpdatedAt": "2022-07-05 15:54:42"
    },
    {
      "Id": 5,
      "Title": "E",
      "CreatedAt": "2022-07-05 15:54:43",
      "UpdatedAt": "2022-07-05 15:54:43"
    },
    {
      "Id": 6,
      "Title": "F",
      "CreatedAt": "2022-07-05 15:54:44",
      "UpdatedAt": "2022-07-05 15:54:44"
    }
  ],
  "pageInfo": {
    "totalRows": 6,
    "page": 1,
    "pageSize": "25",
    "isFirstPage": true,
    "isLastPage": true
  }
}
```

With Shuffling  - Test in Swagger UI (shuffle = 1)

Result:

```json
{
  "list": [
    {
      "Id": 2,
      "Title": "B",
      "CreatedAt": "2022-07-05 15:45:10",
      "UpdatedAt": "2022-07-05 15:45:10"
    },
    {
      "Id": 5,
      "Title": "E",
      "CreatedAt": "2022-07-05 15:54:43",
      "UpdatedAt": "2022-07-05 15:54:43"
    },
    {
      "Id": 6,
      "Title": "F",
      "CreatedAt": "2022-07-05 15:54:44",
      "UpdatedAt": "2022-07-05 15:54:44"
    },
    {
      "Id": 3,
      "Title": "C",
      "CreatedAt": "2022-07-05 15:54:41",
      "UpdatedAt": "2022-07-05 15:54:41"
    },
    {
      "Id": 1,
      "Title": "A",
      "CreatedAt": "2022-07-05 15:45:09",
      "UpdatedAt": "2022-07-05 15:45:09"
    },
    {
      "Id": 4,
      "Title": "D",
      "CreatedAt": "2022-07-05 15:54:42",
      "UpdatedAt": "2022-07-05 15:54:42"
    }
  ],
  "pageInfo": {
    "totalRows": 6,
    "page": 1,
    "pageSize": "25",
    "isFirstPage": true,
    "isLastPage": true
  }
}
```

Result 2:

```json
{
  "list": [
    {
      "Id": 1,
      "Title": "A",
      "CreatedAt": "2022-07-05 15:45:09",
      "UpdatedAt": "2022-07-05 15:45:09"
    },
    {
      "Id": 6,
      "Title": "F",
      "CreatedAt": "2022-07-05 15:54:44",
      "UpdatedAt": "2022-07-05 15:54:44"
    },
    {
      "Id": 2,
      "Title": "B",
      "CreatedAt": "2022-07-05 15:45:10",
      "UpdatedAt": "2022-07-05 15:45:10"
    },
    {
      "Id": 3,
      "Title": "C",
      "CreatedAt": "2022-07-05 15:54:41",
      "UpdatedAt": "2022-07-05 15:54:41"
    },
    {
      "Id": 5,
      "Title": "E",
      "CreatedAt": "2022-07-05 15:54:43",
      "UpdatedAt": "2022-07-05 15:54:43"
    },
    {
      "Id": 4,
      "Title": "D",
      "CreatedAt": "2022-07-05 15:54:42",
      "UpdatedAt": "2022-07-05 15:54:42"
    }
  ],
  "pageInfo": {
    "totalRows": 6,
    "page": 1,
    "pageSize": "25",
    "isFirstPage": true,
    "isLastPage": true
  }
}
```

## Curl 

Without using alias

```
curl --request GET \
  --url 'http://localhost:8080/api/v1/db/data/noco/p_w8d6lhzx4esjg0/Sheet1/views/Sheet1?limit=25&offset=0&where=&shuffle=0' \
  --header 'xc-auth: <REDACTED>'
```

```
curl --request GET \
  --url 'http://localhost:8080/api/v1/db/data/noco/p_w8d6lhzx4esjg0/Sheet1/views/Sheet1?limit=25&offset=0&where=&shuffle=1' \
  --header 'xc-auth: <REDACTED>'
```

Using alias

```
curl --request GET \
  --url 'http://localhost:8080/api/v1/db/data/noco/p_w8d6lhzx4esjg0/Sheet1/views/Sheet1?limit=25&offset=0&where=&r=0' \
  --header 'xc-auth: <REDACTED>'
```

Result

```json
{"list":[{"Id":1,"Title":"A"},{"Id":2,"Title":"B"},{"Id":3,"Title":"C"},{"Id":4,"Title":"D"},{"Id":5,"Title":"E"},{"Id":6,"Title":"F"}],"pageInfo":{"totalRows":6,"page":1,"pageSize":"25","isFirstPage":true,"isLastPage":true}}
```

```
curl --request GET \
  --url 'http://localhost:8080/api/v1/db/data/noco/p_w8d6lhzx4esjg0/Sheet1/views/Sheet1?limit=25&offset=0&where=&r=1' \
  --header 'xc-auth: <REDACTED>'
```

Result 

```json
{"list":[{"Id":4,"Title":"D"},{"Id":3,"Title":"C"},{"Id":1,"Title":"A"},{"Id":5,"Title":"E"},{"Id":6,"Title":"F"},{"Id":2,"Title":"B"}],"pageInfo":{"totalRows":6,"page":1,"pageSize":"25","isFirstPage":true,"isLastPage":true}}
```

Result 2: 

```json
{"list":[{"Id":1,"Title":"A"},{"Id":3,"Title":"C"},{"Id":2,"Title":"B"},{"Id":6,"Title":"F"},{"Id":4,"Title":"D"},{"Id":5,"Title":"E"}],"pageInfo":{"totalRows":6,"page":1,"pageSize":"25","isFirstPage":true,"isLastPage":true}}
```

### CODE REVIEW

**Review by mertmit — COMMENTED** [review]
I have a concern that users might expect us to generate randomized results from all of the query as in related issue `GET  /api/v1/pizzas?random=true&limit=1`
For example if we use example from issue (limit=1) with random we get same results every time.
But if our aim is to sort random within response page, we should state it clearly in documentation too.

**Comment by wingkwong** [comment]
> I have a concern that users might expect us to generate randomized results from all of the query as in related issue `GET /api/v1/pizzas?random=true&limit=1` For example if we use example from issue (limit=1) with random we get same results every time. But if our aim is to sort random within response page, we should state it clearly in documentation too.

Good Catch. I misunderstood the statement.

File: packages/nocodb/src/lib/db/sql-data-mapper/lib/sql/BaseModelSqlv2.ts [hunk 1]
```diff
@@ -852,6 +859,10 @@ class BaseModelSqlv2 {
         .orWhereNull(rcn)
     );
 
+    if (+args?.shuffle) {
+      qb.orderByRaw('RAND()');
```
Comment (mertmit) [inline packages/nocodb/src/lib/db/sql-data-mapper/lib/sql/BaseModelSqlv2.ts:863]: This should use your shuffle function

**Review by mertmit — CHANGES_REQUESTED** [review]

File: packages/nocodb/src/lib/db/sql-data-mapper/lib/sql/BaseModelSqlv2.ts (hunk 1 above)
Comment (wingkwong) [reply packages/nocodb/src/lib/db/sql-data-mapper/lib/sql/BaseModelSqlv2.ts]: Updated.

**Review by mertmit — APPROVED** [review]

**Pull request merged**